---
title: 'Heterozygosity'
editor_options:
  markdown:
    wrap: 72
output:
  html_document:
    df_print: paged
  pdf_document: default
---

Heterozygosity, statistically speaking, is the expected level of heterozygosity in a population or an within an individual at a given locus. We can analyze the average heterozygosity of an individual or look at local heterozygosity throughout a genome.

To analyze heterozygosity, we will do so in the following steps:

1) Average Heterozygosity with bcftools and vcftools

	1a) Calculate heterozygosity
	
	1b) Plot results in R
	
2) Genome-Wide Heterozygosity using ANGSD

  2a) Run ANGSD
  
  2b) Add metadata
  
  2c) Concatenate files
  
  2d) Plot results in R

## 1a) Average Heterozygosity with bcftools and vcftools

Calculate average heterozygosity with bcftools and vcftools. These tools will calculate this statistic for each individual within the .vcf file. Submit the following code on an interactive or computational node:

```Bash
#Replace these with the actual file names
VCF_FILE="/path/to/vcf/file/samples.vcf.gz"
OUTPUT_FILE="NN_heterozygosity_.tsv"
#you can use the fasta index (.fai) to sum the total length of the genome
GENOME_LENGTH=2468345093 

#print list of sample names from the VCF file
SAMPLES=$(bcftools query -l $VCF_FILE)

#Write a header line to the output file
echo -e "Sample\tHeterozygous_sites\tHeterozygosity" > $OUTPUT_FILE

#Loop through each sample and calculate the heterozygosity
for SAMPLE in $SAMPLES; do
  HETEROZYGOUS=$(bcftools view -s $SAMPLE $VCF_FILE | grep -v "#" | grep -o "0/1" | wc -l)
  HETEROZYGOSITY=$(echo "scale=7; $HETEROZYGOUS / $GENOME_LENGTH" | bc)
  echo -e "$SAMPLE\t$HETEROZYGOUS\t$HETEROZYGOSITY" >> $OUTPUT_FILE
done
```
The output should be a tab-separated .tsv file.

## 1b) Plot results in R

Create a scatter plot of heterozygosity values from the resulting .tsv file. You can achieve this plot by running the following code in R:

```
#install ggplot if you haven't done so already
install.packages("tidyverse")
library(ggplot2)

#replace this with the actual file name
INPUT_FILE <- "heterozygosity.tsv"

#read in the data from the input file
data <- read.table(INPUT_FILE, header=TRUE, sep="\t")

#create a scatter plot of heterozygosity values
plot <- ggplot(data, aes(x=Sample, y=Heterozygosity)) +
    geom_point(size=2) +
    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5)) +
    labs(x="Sample", y="Heterozygosity")

#save the plot to a file
ggsave("heterozygosity_plot.png", plot, width=10, height=5, dpi=300)
```

## 2) Genome-Wide Heterozygosity using ANGSD

ANGSD (Analysis of Next Generation Sequencing Data) is a software for analyzing next generation sequencing data. For more information or to download, visit [ANGSD](https://github.com/ANGSD/angsd).

## 2a) Run ANGSD

Be sure modules are available and loaded on your HPC, submit a job with the following code:

```Bash
echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME
 #
#Set variables
input_bam_file="/path/to/file/file.bam"
ancestral_fasta_file="/path/to/reference/genome/file.fasta"
reference_fasta_file="/path/to/reference/genome/file.fasta"
output_directory="/path/where/you/want/output"
SAMPLE="NN114296"

#Loop through scaffolds 1 to 18
for i in {1..18}; do
# Run ANGSD command
angsd -P 10 -i ${input_bam_file} -anc ${ancestral_fasta_file} -dosaf 1 -gl 1 -C 50 -minQ 20 -minmapq 30 -fold 1 -out ${output_directory}/$SAMPLE.SUPER_${i} -ref ${reference_fasta_file} -r SUPER_${i}

# Run realSFS command, Site frequency spectrum describes the distribution of allele frequencies across site in the genome of a particular species
realSFS -nsites 200000 ${output_directory}/$SAMPLE.SUPER_${i}.saf.idx > ${output_directory}/$SAMPLE.SUPER_${i}.est.ml

done
#
#Explanation
#P <threads>: Sets the number of threads to be used in parallel
#i <input_bam_file>: Specifies the input file as a BAM file.
#anc <ancestral_fasta_file>: Specifies the ancestral fasta reference file.
#dosaf <dosaf_value>: Computes the Site Frequency Spectrum (SFS) based on the genotype likelihoods
#gl <genotype_likelihood_method>: Specifies the method used for calculating genotype likelihoods
#C <base_quality_adjustment>: Adjusts the base quality score by a specified value before using it
#minQ <min_base_quality>: Sets the minimum base quality score required
#minmapq <min_mapping_quality>:Sets the minimum mapping quality score required
#fold <fold_value>:Indicates whether you are analyzing folded SFS or unfolded SFS
#out <output_file>: Specifies the output file path and name.
#ref <reference_fasta_file>: Specifies the reference fasta file
#r <region_of_interest>: Specifies the region of interest, note that ANGSD cannot differentiate chromosomes if the whole genome is run at once
#nsites <number_of_sites>: Specifies the number of sites to be considered for the estimation. Replace <number_of_sites> with the desired number.
#
echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME
```

## 2b) Add metadata

Add sample name and scaffold number for each line. Submit the following code on an interactive or computational node:

```Bash
output_directory="path/to/your/output_directory"
SAMPLE="your_sample_name"

# Loop through scaffolds 1 to 18
for i in {1..18}; do
  # Calculate the number of lines in the output file
  num_lines=$(wc -l < ${output_directory}/$SAMPLE.SUPER_${i}.est.ml)

  # Add the number of lines, sample name, and scaffold number to each line of the output file
  awk -v lines="$num_lines" -v sample="$SAMPLE" -v scaffold="$i" '{print lines, sample, "SUPER_" scaffold, $0}' ${output_directory}/$SAMPLE.SUPER_${i}.est.ml > ${output_directory}/$SAMPLE.SUPER_${i}.est.ml.annotated

  # Optional: Move the annotated file to the original file
  mv ${output_directory}/$SAMPLE.SUPER_${i}.est.ml.annotated ${output_directory}/$SAMPLE.SUPER_${i}.est.ml

done
```

## 2c) Concatenate files

Concatenate files by submitting the following code on an interactive or computational node:

```Bash
#Set the input and output directory
input_directory="path/to/your/input_directory"
output_directory="path/to/your/output_directory"

#Create an output file
output_file="${output_directory}/all_est_ml_concatenated.txt"

#Remove the output file if it already exists
if [ -f "${output_file}" ]; then
rm "${output_file}"
fi

#Loop through all "[est.ml](http://est.ml/)" files and concatenate them
for file in ${input_directory}/*.est.ml; do
cat "${file}" >> "${output_file}"

done
```

## 2d) Plot results in R

Use the following code to plot your results in R:

```R
#Load required libraries
library(tidyverse) #for data manipulation and visualization
library(viridis)   #for generating color palettes
library(scales)    #for scaling and formatting axis labels

# Read data file and store it in the variable 'het_master'
het_master <- read.table("/path/to/file/file.est.ml")

# Manipulate data
het_master %>% 
  rename(sample=V2) %>%          # Rename V2 as 'sample'
  rename(chromosome = V3) %>%    # Rename V3 as 'chromosome'
  mutate(heterozygosity = V5/(V4 + V5)) %>% # Calculate heterozygosity as V5 / (V4 + V5)
  mutate(position = ((V1*200000)-200000))   %>% # Calculate position as (V1 * 200000) - 200000
  filter(chromosome == "SUPER_2") %>% # Filter data to include only rows where 'chromosome' is 'SUPER_2'

# Create a ggplot2 plot
  ggplot(aes(x=position, y=heterozygosity)) + # Set x-axis as 'position' and y-axis as 'heterozygosity'
  geom_line(colour="grey",alpha=0.5) + # Add a line plot with grey color and 0.5 alpha (transparency)
  geom_point(aes(colour=factor(chromosome))) + # Add points, color them based on 'chromosome' factor variable
  scale_color_viridis(discrete = TRUE) + # Use viridis color palette for the points
  facet_grid(sample ~ chromosome,scales = "free_x") + # Create a facet grid with 'sample' on the y-axis and 'chromosome' on the x-axis, set x-axis scales to be free
  labs(x = NULL, y = "Heterozygosity\n") + # Remove x-axis labels and set y-axis label to "Heterozygosity\n"
  scale_y_continuous(labels = comma) + # Format y-axis labels with commas
  scale_x_continuous(labels = comma) + # Format x-axis labels with commas
  theme_minimal() + # Apply a minimal theme to the plot
  theme(legend.position = "none", # Remove legend
        strip.text.x = element_text(face = "bold"), # Set strip text for x-axis to bold
        strip.text.y = element_text(face = "bold"), # Set strip text for y-axis to bold
        panel.grid.major.x = element_blank(), # Remove major x-axis gridlines
        panel.grid.minor.x = element_blank(), # Remove minor x-axis gridlines
        panel.spacing.x = unit(0, "line"), # Set panel spacing to zero
        panel.border = element_rect(color = "black", fill = NA, size = 0.25)) # Add a black border around the panels
```

